bpf/bpf_overlay.c

```diff
@@ -305,8 +312,19 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx, __u32 *identity)
 		if (ep->flags & ENDPOINT_F_HOST)
 			goto to_host;
 
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE && \
+	!defined(TUNNEL_MODE) && !defined(ENABLE_HOST_ROUTING)
+		/* In conjunction with the direct routing mode, DSR with Geneve dispatch
+		 * and the legacy routing mode, we pass packets to the kernel stack.
+		 * Otherwise, the kernel stack will only be aware of NON-SYN packets
+		 * because we encapsulate SYN packet only in DSR with Geneve dispatch.
+		 */
+		ctx_change_type(ctx, PACKET_HOST);
+		return CTX_ACT_OK;
+#else
 		return ipv4_local_delivery(ctx, ETH_HLEN, *identity, ip4, ep,
 					   METRIC_INGRESS, false, false);
+#endif
 	}
 
```

Comment:
I have a feeling we should/need to integrate this into `ipv4_local_delivery()` better - passing a `deliver-via-kernel` flag or such.

Also need to be careful about interactions with Egress-Gateway reply traffic when in Direct-Routing mode ...

Reply:
> Also need to be careful about interactions with Egress-Gateway reply traffic when in Direct-Routing mode ...

Ah, right, I will check the interaction with Egress GW

Reply:
I'm still a bit stuck on this part - primarily because it changes behaviour for EgressGW reply traffic, as it passes through `from-overlay` on the worker node. Can you sketch out a bit more why the change needed, and how come we didn't need it for EgressGW traffic before?

Reply:
This change is intended primarily for the generic veth chain use case. In the chaining mode, the other chained plugin is responsible for the underlying connectivity. I thought that bypassing the kernel stack here would break this assumption.
https://github.com/cilium/cilium/blob/faa23a078767f1892325f15fceaa7b2fc14f3ca8/install/kubernetes/cilium/templates/cilium-configmap.yaml#L309-L313

EgressGW also supports the generic veth chain mode, right? If so, shouldn't we do this in the first place?

Reply:
As far as I looked into the packet flow under the CNI chain mode, we pass packets to the kernel stack here and it delivers them to the local endpoint.
https://github.com/cilium/cilium/blob/d56653d6f63557088c4c5065ff9cf2796777df9a/bpf/bpf_host.c#L534

Reply:
Dropped the deliver via kernel commit for now. Let's discuss it as a follow-on

Reply:
Ack, thanks!